src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,101 +3,356 @@
+    }
+
+    private void WriteSentinelByteToPipeWriter()
+    {
+        var span = _pipeWriter.GetSpan(1);
+        span[0] = MessageSentinelByte;
+        _pipeWriter.Advance(1);
+    }
+
+    private void WriteTempStreamLengthToPipeWriter(ReadWriteStream tempStream)
+    {
+        var length = tempStream.Length;
+        Contract.ThrowIfTrue(length > int.MaxValue);
+
+        var span = _pipeWriter.GetSpan(sizeof(int));
+        BinaryPrimitives.WriteInt32LittleEndian(span, (int)length);
+        _pipeWriter.Advance(sizeof(int));
+    }
+}
+
+internal readonly struct RemoteHostAssetReader<T, TArg>(
+    PipeReader pipeReader,
+    Checksum solutionChecksum,
+    int objectCount,
+    ISerializerService serializer,
+    Action<Checksum, T, TArg> callback,
+    TArg arg)
+{
+    private readonly PipeReader _pipeReader = pipeReader;
+    private readonly Checksum _solutionChecksum = solutionChecksum;
+    private readonly int _objectCount = objectCount;
+    private readonly ISerializerService _serializer = serializer;
+    private readonly Action<Checksum, T, TArg> _callback = callback;
+    private readonly TArg _arg = arg;
+
+    public ValueTask ReadDataAsync(CancellationToken cancellationToken)
+    {
+        // Suppress ExecutionContext flow for asynchronous operations operate on the pipe. In addition to avoiding
+        // ExecutionContext allocations, this clears the LogicalCallContext and avoids the need to clone data set by
+        // CallContext.LogicalSetData at each yielding await in the task tree.
+        //
+        // ⚠ DO NOT AWAIT INSIDE THE USING. The Dispose method that restores ExecutionContext flow must run on the
+        // same thread where SuppressFlow was originally run.
+        using var _ = FlowControlHelper.TrySuppressFlow();
+        return ReadDataSuppressedFlowAsync(cancellationToken);
+    }
+
+    private async ValueTask ReadDataSuppressedFlowAsync(CancellationToken cancellationToken)
+    {
+        using var pipeReaderStream = _pipeReader.AsStream(leaveOpen: true);
+
+        // Get an object reader over the stream.  Note: we do not check the validation bytes here as the stream is
+        // currently pointing at header data prior to the object data.  Instead, we will check the validation bytes
+        // prior to reading each asset out.
+        using var objectReader = ObjectReader.GetReader(pipeReaderStream, leaveOpen: true, checkValidationBytes: false);
+
+        // Ensure that no invariants were broken and that both sides of the communication channel are talking about
+        // the same pinned solution.
+        var responseSolutionChecksum = await ReadChecksumFromPipeReaderAsync(cancellationToken).ConfigureAwait(false);
+        Contract.ThrowIfFalse(_solutionChecksum == responseSolutionChecksum);
+
+        // Now actually read all the messages we expect to get.
+        for (int i = 0, n = _objectCount; i < n; i++)
```

Comment:
>n = _objectCount [](http://example.com/codeflow?start=24&length=16)

curious: is it somehow better to compare to a local rather than the field?

Reply:
not really no.  this is just how the original loop was written.  so i kept as is.  i'll update.